File: hw3/scala/task2train.scala
import org.apache.spark.storage.StorageLevel.{MEMORY_AND_DISK_SER,DISK_ONLY}
import org.apache.spark.{SparkConf, SparkContext}
import org.json4s.jackson.JsonMethods.{compact, parse}

import java.io.{File, PrintWriter}
import java.util.Date
import scala.math.log10
import scala.util.control.Exception.allCatch

object task2train {

  def is_number(s: String): Boolean = (allCatch opt s.toDouble).isDefined
  def remove_words(w_list: Iterable[String], stop_words: List[String]) = {
    var new_words:List[String] = List()
    for (elem <- w_list) {
      if (!stop_words.contains(elem) && !is_number(elem)) {
        new_words = new_words :+ elem
      }
    }
    new_words
  }

  def main(args: Array[String]): Unit = {
    val start_time = new Date().getTime
    val conf = new SparkConf().setMaster("local[*]").setAppName("task1")
    val sc = new SparkContext(conf)
//    val train_file = "E:\\Data\\inf553\\hw3\\publicdata\\train_review.json"
//    val model_file = "task2.scala.model"
//    val stopwords = "E:\\Data\\inf553\\hw3\\publicdata\\stopwords"
    val train_file = args(0)
    val model_file = args(1)
    val stopwords = args(2)

    val rdd_pre = sc.textFile(train_file).map(item => parse(item)).map(item => (compact(item \ "business_id"), compact(item \ "user_id"), compact(item \ "text")))
//    val business_index = rdd_pre.map(_._1).zipWithIndex().collectAsMap()
//    val user_index = rdd_pre.map(_._2).zipWithIndex().collectAsMap()
    val rdd = rdd_pre.map(item => (item._1, item._3)).groupByKey()
      .mapValues(item=>item.flatMap(x=>x.replaceAll("[(|)|\\[|\\]|\"|,|.|!|:|;|?]","").split(" ")))


    val stop_words = sc.textFile(stopwords).collect().toList

    val rdd1 = rdd.mapValues(item=>item.map(_.toLowerCase().trim)).mapValues(item=>item.filter(_ != "")).mapValues(item=>item.filter(i=>(!stop_words.contains(i) && !is_number(i))))
//      .mapValues(item=>item.filter(!is_number(_)))

    val total_num_words = rdd1.map(item=>item._2.size).reduce((x, y)=> x + y)
    val threshold = total_num_words * 0.000001

    val no_rare_words_index = rdd1.flatMap(item=>item._2.map((_, 1))).reduceByKey((x,y)=>x+y).filter(item=>item._2 >= threshold).map(_._1).zipWithIndex().collectAsMap()
    val rdd2 = rdd1.mapValues(item=>item.filter(x=>no_rare_words_index.contains(x)).map(no_rare_words_index.get(_).head.toInt))//.persist(MEMORY_AND_DISK_SER)
    val n_business = rdd.count()
    val idf_dict = rdd2.flatMap(_._2.toSet).map((_, 1)).groupByKey().mapValues(_.sum).map(item=>(item._1, log10(n_business/item._2))).collect().toMap
    val rdd_tf_idf = rdd2.mapValues(item=>item.map((_,1.0))).mapValues(item=>item.groupBy(_._1).mapValues(_.size).toList)
      .mapValues(item=>{
        val max_f = item.map(_._2).max
        val r = item.map(i=>(i._1,i._2 * 1.0/max_f*idf_dict.get(i._1).head)).sortBy(_._2).reverse.map(_._1)
        if(r.size >= 200) {
          r.take(200)
        }else{
          r
        }
      })
    val business_profile_dict = rdd_tf_idf.collect().toMap

    val rdd_user_profile = rdd_pre.map(item => (item._2, item._1)).groupByKey().mapValues(item=> {
      var r: List[Int] = List()
      item.foreach(i=>r = r++business_profile_dict.get(i).head)
      r.distinct
    })

    val pw = new PrintWriter(new File(model_file))
    rdd_tf_idf.collect().foreach(item=>pw.write("{\"type\": \"b\", \"business_id\": "+item._1 + ", \"business_profile\": " +item._2.map(_.toString).mkString("[", ",", "]") + "}\n"))
    rdd_user_profile.collect().foreach(item=>pw.write("{\"type\": \"u\", \"user_id\": "+item._1 + ", \"user_profile\": " +item._2.map(_.toString).mkString("[", ",", "]") + "}\n"))
    pw.close()
    val end_time =new Date().getTime
    println("time: "+(end_time-start_time)/1000.0+"s")
  }
}

File: hw3/python/task2train.py
import math
import os
import sys
import json
import time
from pyspark import SparkContext

def is_number(s):
    try:
        float(s)
        return True
    except ValueError:
        pass
    try:
        import unicodedata
        unicodedata.numeric(s)
        return True
    except (TypeError, ValueError):
        pass
    return False

def remove_words(w_list, stop_words):
    new_words = list()
    for w in w_list:
        if w not in stop_words and not is_number(w):
            new_words.append(w)
    return new_words

def filter_rare(w_list, no_rare_words_index):
    new_words = list()
    # print(w_list)
    for w in w_list:
        index = no_rare_words_index.get(w)
        if index is not None:
            new_words.append(index)
    return new_words

def calculate_tf_idf(x, idf_dict):
    business, word_list = x[0], list(x[1])
    new_words_dict = dict()
    for w in word_list:
        new_words_dict[w] = new_words_dict.get(w, 0) + 1
    # calculate tf-idf
    max_f = 0
    for k,v in new_words_dict.items():
        max_f = v if v > max_f else max_f
    for k in new_words_dict.keys():
        new_words_dict[k] = new_words_dict[k]/max_f * idf_dict[k]
    a = sorted(new_words_dict.items(), key=lambda x: x[1], reverse=True)
    words_dict = dict(a[:200]) if len(a) > 200 else dict(a)
    return business, list(words_dict.keys())

def aggregate_item(business_ids, business_vector_dict):
    tmp = list()
    for id in business_ids:
        tmp += business_vector_dict[id]
    return list(set(tmp))

if __name__ == '__main__':

    start_time = time.time()
    os.environ['JAVA_HOME'] = '/home/buyi/app/jdk'
    os.environ['PYSPARK_PYTHON'] = '/usr/bin/python3'
    train_file, model_file, stopwords = '/home/buyi/data/hw3/train_review.json', 'task2.model', '/home/buyi/data/hw3/stopwords'
    # train_file, model_file, stopwords = sys.argv[1:]

    sc = SparkContext('local[*]', 'task2')
    rdd_review = sc.textFile(train_file).map(json.loads).map(lambda x: (x['business_id'], x['text']))
    # 1. concatenate all reviews for a business
    rdd = rdd_review.groupByKey().mapValues(lambda x:" ".join(list(x))).mapValues(lambda x: [i.strip().lower() for i in x.split(' ')])
    punctuations = ["(", "[", ",", ".", "!", "?", ":", ";", "]", ")"]
    stop_words = sc.textFile(stopwords).collect()
    stop_words.append("")
    for p in punctuations:
        stop_words.append(p)
    # print(rdd.take(1))
    # print("the number of stop words: {}".format(len(stop_words)))
    rdd1 = rdd.mapValues(lambda x: remove_words(x, stop_words))

    total_num_words = rdd1.map(lambda x: len(x[1])).reduce(lambda x, y: x + y)
    threshold = total_num_words * 0.000001
    no_rare_words_index = dict(rdd1.flatMap(lambda x: [(i, 1) for i in x[1]]).reduceByKey(lambda x, y: x + y).filter(
        lambda x: x[1] >= threshold).map(lambda x: x[0]).zipWithIndex().collect())
    print("the number of not rare words: {}".format(len(no_rare_words_index)))
    rdd2 = rdd1.mapValues(lambda x: filter_rare(x, no_rare_words_index))

    n_business = rdd.count()
    print("number of business: {}".format(n_business))
    rdd_idf = rdd2.flatMap(lambda x:set(x[1])).map(lambda x:(x, 1)).groupByKey().mapValues(lambda x:sum(x)).map(lambda x: (x[0], math.log(n_business/ x[1], 10)))
    idf_dict = dict(rdd_idf.collect())
    print("time: {}".format(time.time() - start_time))

    rdd_tf_idf = rdd2.map(lambda x: calculate_tf_idf(x, idf_dict))
    business_profile_dict = dict(rdd_tf_idf.collect())
    print("time: {}".format(time.time() - start_time))

    rdd3 = sc.textFile(train_file).map(json.loads).map(lambda x: (x['user_id'], x['business_id']))
    rdd_user_profile = rdd3.groupByKey().mapValues(lambda x:aggregate_item(x,business_profile_dict))
    user_profile_dict = dict(rdd_user_profile.collect())
    # print(user_profile_dict)


    with open(model_file, 'w') as f:
        f.truncate()
        for k,v in business_profile_dict.items():
            f.write(json.dumps({"type":"b", "business_id":k, "business_profile": v})+'\n')
        for k,v in user_profile_dict.items():
            f.write(json.dumps({"type":"u", "user_id":k, "user_profile":v})+'\n')

    print("time: {}".format(time.time() - start_time))